controller/openebs/cstor.go

```diff
@@ -50,5 +61,199 @@ func (p *Planner) setCStorDefaultsIfNotSet() error {
+		p.ObservedOpenEBS.Spec.CstorConfig.CVCOperator.ImageTag = p.ObservedOpenEBS.Spec.Version
+	}
+	// form the container image as per the image prefix and image tag.
+	p.ObservedOpenEBS.Spec.CstorConfig.CVCOperator.Image = p.ObservedOpenEBS.Spec.ImagePrefix +
+		"cvc-operator:" + p.ObservedOpenEBS.Spec.CstorConfig.CVCOperator.ImageTag
+	if p.ObservedOpenEBS.Spec.CstorConfig.CVCOperator.Replicas == nil {
+		p.ObservedOpenEBS.Spec.CstorConfig.CVCOperator.Replicas = new(int32)
+		*p.ObservedOpenEBS.Spec.CstorConfig.CVCOperator.Replicas = DefaultCVCOperatorReplicaCount
+	}
+	return nil
+}
+
+// updateCSPCOperator updates the CSPC operator manifest as per the reconcile.ObservedOpenEBS values.
+func (p *Planner) updateCSPCOperator(deploy *unstructured.Unstructured) error {
+	// get the containers of the cspc-operator and update the desired fields
+	containers, err := unstruct.GetNestedSliceOrError(deploy, "spec", "template", "spec", "containers")
+	if err != nil {
+		return err
+	}
+	// update the env value of cspc-operator container
+	updateCSPCOperatorEnv := func(env *unstructured.Unstructured) error {
+		envName, _, err := unstructured.NestedString(env.Object, "spec", "name")
+		if err != nil {
+			return err
+		}
+		if envName == "OPENEBS_IO_BASE_DIR" {
+			err = unstructured.SetNestedField(env.Object, p.ObservedOpenEBS.Spec.DefaultStoragePath,
+				"spec", "value")
+		} else if envName == "OPENEBS_IO_CSTOR_POOL_SPARSE_DIR" {
+			err = unstructured.SetNestedField(env.Object,
+				p.ObservedOpenEBS.Spec.DefaultStoragePath+"/sparse", "spec", "value")
+		} else if envName == "OPENEBS_IO_CSPI_MGMT_IMAGE" {
+			err = unstructured.SetNestedField(env.Object,
+				p.ObservedOpenEBS.Spec.CstorConfig.CSPIMgmt.Image, "spec", "value")
+		} else if envName == "OPENEBS_IO_CSTOR_POOL_IMAGE" {
+			err = unstructured.SetNestedField(env.Object,
+				p.ObservedOpenEBS.Spec.CstorConfig.Pool.Image, "spec", "value")
+		} else if envName == "OPENEBS_IO_CSTOR_POOL_EXPORTER_IMAGE" {
+			err = unstructured.SetNestedField(env.Object,
+				p.ObservedOpenEBS.Spec.Policies.Monitoring.Image, "spec", "value")
+		}
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+	updateContainer := func(obj *unstructured.Unstructured) error {
+		containerName, _, err := unstructured.NestedString(obj.Object, "spec", "name")
+		if err != nil {
+			return err
+		}
+		envs, _, err := unstruct.GetSlice(obj, "spec", "env")
+		if err != nil {
+			return err
+		}
+		// update the envs of cspc-operator container
+		// In order to update envs of other containers, just write an updateEnv
+		// function for specific containers.
+		if containerName == "cspc-operator" {
+			err = unstruct.SliceIterator(envs).ForEachUpdate(updateCSPCOperatorEnv)
+			if err != nil {
+				return err
+			}
+		}
+		err = unstructured.SetNestedSlice(obj.Object, envs, "spec", "env")
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	err = unstruct.SliceIterator(containers).ForEachUpdate(updateContainer)
```

Comment:
n_: can we use same package
unstructured -> unstruct

Reply:
This is unstruct only, right?

Reply:
sorry wrong place